src/saleor_app/security.py

```diff
@@ -0,0 +1,174 @@
+import hashlib
+import hmac
+import logging
+from typing import Optional
+
+import jwt
+from fastapi import HTTPException, Request, Security
+from fastapi.openapi.models import APIKey as APIKeyModel
+from fastapi.openapi.models import APIKeyIn
+from fastapi.security.base import SecurityBase
+from jwt.api_jwk import PyJWKSet
+from pydantic import BaseModel
+from starlette.status import HTTP_401_UNAUTHORIZED
+
+from saleor_app.schemas.core import Saleor
+from saleor_app.schemas.principals import SaleorPrincipal
+
+logger = logging.getLogger(__name__)
+
+SALEOR_DOMAIN_HEADER = "x-saleor-domain"
+SALEOR_TOKEN_HEADER = "x-saleor-token"
+SALEOR_SIGNATURE_HEADER = "x-saleor-signature"
+
+
+class SaleorDomainSecurity(SecurityBase):
+    def __init__(
+        self,
+        *,
+        auto_error: bool = True,
+    ):
+        self.model = APIKeyModel(
+            **{"in": APIKeyIn.header},
+            name=SALEOR_DOMAIN_HEADER,
+            description="Uses app's validate_domain method to check if a Saleor domain is authorized to use this app.",
+        )
+        self.scheme_name = "Saleor Domain Validation"
+        self.auto_error = auto_error
+
+    async def __call__(
+        self,
+        request: Request,
+    ) -> Optional[Saleor]:
+
+        saleor: str = request.headers.get(SALEOR_DOMAIN_HEADER)
+
+        domain_is_valid = await request.app.validate_domain(saleor)
+        if not domain_is_valid and self.auto_error:
+            raise HTTPException(
+                status_code=HTTP_401_UNAUTHORIZED,
+                detail=f"Provided domain {saleor} is invalid.",
+            )
+        return Saleor(domain=saleor)
+
+
+class SaleorUserAuthSecurity(SecurityBase):
+    def __init__(
+        self,
+        *,
+        auto_error: bool = True,
+    ):
+        self.model = APIKeyModel(
+            **{"in": APIKeyIn.header},
+            name=SALEOR_TOKEN_HEADER,
+            description="Authenticates the user with the JWT issued by Saleor.",
+        )
+        self.scheme_name = "Saleor User Authentication"
+        self.auto_error = auto_error
+
+    async def __call__(
+        self,
+        request: Request,
+        saleor: Saleor = Security(SaleorDomainSecurity(auto_error=False)),
+    ) -> Optional[SaleorPrincipal]:
+
+        if not saleor:
+            if self.auto_error:
+                raise HTTPException(
+                    status_code=HTTP_401_UNAUTHORIZED,
+                    detail=f"Provided domain {saleor} is invalid.",
+                )
+            else:
+                return None
+
+        token: str = request.headers.get(SALEOR_TOKEN_HEADER)
+
+        if not token:
+            logger.warning(f"Missing {SALEOR_TOKEN_HEADER.upper()} header.")
+            if self.auto_error:
+                raise HTTPException(
+                    status_code=HTTP_401_UNAUTHORIZED,
+                    detail=f"Missing {SALEOR_TOKEN_HEADER.upper()} header.",
```

Comment:
Can we be consistent across errors?
```suggestion
                    detail=f"Missing token {SALEOR_TOKEN_HEADER.upper()} header.",
```
Maybe create a custom exception class that would be consistent with generating error messages?

Reply:
We could do that to allow some control over the error payload of a service which would be good but might the detrimental - now if an app does not care about a custom payload you do not need to do anything. If it does want to hide this default and use a different error schema it would be a matter of making an error handler for `HTTPException`. 
If we went for a custom exception here we would force the user to create a custom error handler every time.